src/runtime-rs/crates/kata/src/task_service.rs

```diff
@@ -0,0 +1,874 @@
+                format!("Failed to resume container {} with error {:?}", cid, e),
+            )
+        })?;
+        Ok(Empty::new())
+    }
+
+    fn do_kill(&self, req: shim::KillRequest) -> Result<Empty> {
+        let cid = req.get_id();
+        let all = req.get_all();
+        let eid = req.get_exec_id();
+        let signal: Signal = Signal::try_from(req.get_signal() as i32).map_err(|e| {
+            get_rpc_status(
+                Code::INTERNAL,
+                format!(
+                    "Failed to get kill signal {} with error: {:?}",
+                    req.get_signal(),
+                    e
+                ),
+            )
+        })?;
+
+        let mut sandbox = self.sandbox.lock().expect("poisoned sandbox lock");
+
+        sandbox.signal_process(cid, eid, signal, all).map_err(|e| {
+            get_rpc_status(
+                Code::INTERNAL,
+                format!(
+                    "Failed to Kill Container {} process {} by Signal {} with error: {:?}",
+                    cid, eid, signal, e
+                ),
+            )
+        })?;
+
+        Ok(Empty::new())
+    }
+
+    fn do_exec(&self, req: shim::ExecProcessRequest) -> Result<Empty> {
+        let cid = req.get_id();
+        let exec_id = req.get_exec_id();
+        let stdin = (!req.stdin.is_empty()).then(|| req.stdin.clone());
+        let stdout = (!req.stdout.is_empty()).then(|| req.stdout.clone());
+        let stderr = (!req.stderr.is_empty()).then(|| req.stderr.clone());
+        let terminal = req.get_terminal();
+
+        //FIXME
+        if req.get_spec().get_type_url().is_empty() {
+            return Err(get_rpc_status(
+                Code::INVALID_ARGUMENT,
+                "Failed to exec Container: Invalid argument",
+            ));
+        }
+
+        let p = serde_json::from_slice::<oci_spec::runtime::Process>(req.get_spec().get_value()).map_err(|e| {
+                get_rpc_status(
+                    Code::INTERNAL,
+                    format!("Failed to deserialize process for container {} process {} with error: {:?}", cid, exec_id, e),
+                )
+            })?;
+
+        let mut sandbox = self.sandbox.lock().expect("poisoned sandbox lock");
+        sandbox
+            .create_exec(cid, exec_id, stdin, stdout, stderr, terminal, p)
+            .map_err(|e| {
+                get_rpc_status(
+                    Code::INTERNAL,
+                    format!(
+                        "Failed to exec container {} process {} with error: {:?}",
+                        cid, exec_id, e
+                    ),
+                )
+            })?;
+        Ok(Empty::new())
+    }
+
+    fn do_resize_pty(&self, req: shim::ResizePtyRequest) -> Result<Empty> {
+        let cid = req.get_id();
+        let exec_id = req.get_exec_id();
+        let height = req.get_height();
+        let width = req.get_width();
+
+        let mut sandbox = self.sandbox.lock().expect("poisoned sandbox lock");
```

Comment:
I wonder if it's worth creating a `get_sandbox!()` macro like `get_controller_or_return_singular_none()` to do this in every function?